salt/iiif/init.sls

```diff
@@ -0,0 +1,63 @@
+# neccessary for docker_* states
+# this is like the python mysql library for mysql_* Salt states
+docker-py:
+    cmd.run:
+        - name: python3 -m pip install docker==4.1.0
+
+loris-repository:
+    git.latest:
+        - name: https://github.com/elifesciences/loris-docker
+        - rev: elife
+        - branch: elife
+        - force_fetch: True
+        - force_checkout: True
+        - force_reset: True
+        - target: /opt/loris-docker
+
+loris-config:
+    file.managed:
+        - name: /opt/loris-docker/loris2.conf
+        - source: salt://iiif/config/etc-loris2-loris2.conf
+        - template: jinja
+        - require:
+            - loris-repository
+
+loris-uwsgi-config:
+    file.managed:
+        # systemd service file expects to find uwsgi.ini in app folder
+        # see builder-base.uwsgi
+        - name: /opt/loris-docker/uwsgi.ini
+        - source: salt://iiif/config/etc-loris2-uwsgi.ini
+        - template: jinja
+        - require:
+            - loris-repository
+
+loris-wsgi-entry-point:
+    file.managed:
+        - name: /opt/loris-docker/loris2.wsgi
+        - source: salt://iiif/config/var-www-loris2-loris2.wsgi
+        - require:
+            - loris-repository
+
+# TODO: these are accumulating
+build-loris:    
+    docker_image.present:
+        - name: elifesciences/loris
+        - build: /opt/loris-docker/
```

Comment:
building the image in every EC2 instance can cause trouble due to unreproducible builds, random errors, etc; I'd strongly suggest to build it in the Jenkinsfile and push it to https://hub.docker.com/r/elifesciences/ so that here you just need to pull and run it skipping the minutes it would take to rebuild.

Reply:
That was my intention once I have everything configured and working.

Because the configuration may change and the configuration needs to live within the container, is there a way to get that config in there without including it as part of the build? I suspect there is, I'll investigate

Reply:
> what I'm doing is having Salt generate the config files as per usual, then mounting/binding these files present on the host to their location within the container.

This is the way to go - one immutable image running everywhere with volumes that mount configuration or data. Volumes is a misleading name because it suggests disks, but it can easily be single files in `/etc`.

Reply:
When we have multiple containers we use `docker-compose` to [organize them all](https://github.com/elifesciences/journal-formula/blob/master/salt/journal/config/srv-journal-docker-compose.yml), start them in the right order, having a single command that acts on all of them.